app/controllers/application_controller.rb

```diff
@@ -1,10 +1,84 @@
 require './config/environment'
-
+require 'pry'
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do 
+    erb :index
+  end
+
+  get '/signup' do
+    # binding.pry
+    if !logged_in 
+      erb :signup
+    else
+     redirect '/tweets'
+    end
+  end
+
+  post '/signup' do
+    if valid_new_user(params)
+      @user = User.new(username: params[:username],email: params[:email],password_digest: BCrypt::Password.create(params[:password]))
+      @user.save
+      session[:id] = @user.id
+      # binding.pry
+      redirect '/tweets'
+    else
+      # erb :'/signup', locals: {message: "Enter valid information."}
+      redirect '/signup'
+    end
+  end 
+
+  get '/login' do 
+    erb :login
   end
 
+  post '/login' do 
+    user = User.find_by(username: params[:username])
+    # binding.pry
+    if user && user.authenticate(params[:password])
+      # binding.pry
+      session[:id] = user.id
+      redirect '/tweets'
+    else
+      redirect "/login"
+    end
+  end
+
+  get '/tweets' do 
+    erb :tweets
+  end
+
+  get '/logout' do 
+    session.clear
+    redirect '/'
+  end
+
+  helpers do 
+    def valid_new_user(params)
```

Comment:
I think you could probably rely on the ActiveRecord model validations instead of buidling a custom validator, e.g. `validates :username, presence: true, uniqueness: true`.